Sources/GRPC/GRPCClientChannelHandler.swift

```diff
@@ -0,0 +1,374 @@
+
+  /// Creates a new gRPC channel handler for clients to translate HTTP/2 frames to gRPC messages.
+  ///
+  /// - Parameters:
+  ///   - streamID: The ID of the HTTP/2 stream that this handler will read and write HTTP/2
+  ///     frames on.
+  ///   - callType: Type of RPC call being made.
+  ///   - logger: Logger.
+  public init(streamID: HTTP2StreamID, callType: GRPCCallType, logger: Logger) {
+    self.streamID = streamID
+    self.logger = logger
+    switch callType {
+    case .unary:
+      self.stateMachine = .init(requestArity: .one, responseArity: .one, logger: logger)
+    case .clientStreaming:
+      self.stateMachine = .init(requestArity: .many, responseArity: .one, logger: logger)
+    case .serverStreaming:
+      self.stateMachine = .init(requestArity: .one, responseArity: .many, logger: logger)
+    case .bidirectionalStreaming:
+      self.stateMachine = .init(requestArity: .many, responseArity: .many, logger: logger)
+    }
+  }
+}
+
+// MARK: - GRPCClientChannelHandler: Inbound
+extension GRPCClientChannelHandler: ChannelInboundHandler {
+  public typealias InboundIn = HTTP2Frame
+  public typealias InboundOut = GRPCClientResponsePart2<Response>
+
+  public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    let frame = self.unwrapInboundIn(data)
+    switch frame.payload {
+    case .headers(let content):
+      self.readHeaders(content: content, context: context)
+
+    case .data(let content):
+      self.readData(content: content, context: context)
+
+    // We don't need to handle other frame type, just drop them instead.
+    default:
+      // TODO: synthesise a more precise `GRPCStatus` from RST_STREAM frames in accordance
+      // with: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#errors
+      break
+    }
+  }
+
+  /// Read the content from an HTTP/2 HEADERS frame received from the server.
+  ///
+  /// We can receive headers in two cases:
+  /// - when the RPC is being acknowledged, and
+  /// - when the RPC is being terminated.
+  ///
+  /// It is also possible for the RPC to be acknowledged and terminated at the same time, the
+  /// specification refers to this as a "Trailers-Only" response.
+  ///
+  /// - Parameter content: Content of the headers frame.
+  /// - Parameter context: Channel handler context.
+  private func readHeaders(content: HTTP2Frame.FramePayload.Headers, context: ChannelHandlerContext) {
+    // In the case of a "Trailers-Only" response there's no guarantee that end-of-stream will be set
+    // on the headers frame: end stream may be sent on an empty data frame as well. If the headers
+    // contain a gRPC status code then they must be for a "Trailers-Only" response.
+    if content.endStream || content.headers.contains(name: GRPCHeaderName.statusCode) {
+      // We have the headers, pass them to the next handler:
+      context.fireChannelRead(self.wrapInboundOut(.trailingMetadata(content.headers)))
```

Comment:
Not quite sure; it just feels weird that we send the HTTP trailers _and_ the gRPC status (extracted from those trailers) in separate "packets".

Reply:
I think this boils down to: "should the user be able to access the trailers if a valid `GRPCStatus` could _not_ be extracted from the trailers?"

I can't think of any compelling reason to not allow the user to access them if we couldn't extract a status.

Reply:
I agree that the user should be able to access the trailers even if we can't extract a status. However, should we maybe just synthesize an "unable to extract status `internalError` (or some other kind of protocol error)" in that case and pass that to the user, along with the other trailers?

Reply:
That's what will happen now, but the we'll fire an error and the status will be synthesised in the handler which catches it.

Reply:
Not sure I understand what the difference is, but I guess it is okay the way it is.

Reply:
This channel handler is not responsible for closing the channel, only for translating HTTP2 frames into our gRPC client message parts. 

Downstream handlers are responsible for surfacing those messages to the user as well as handling any errors (since there could be promises which need to be failed). This might also include logging them or calling out to a delegate and so on. 

The difference between handling the error here vs. in a downstream handler is just consistency with the contract we set out for this handler.

Reply:
👍 Thank you for the explanation!